truffleruby/src/main/java/org/truffleruby/language/dispatch/CachedDispatchNode.java

```diff
@@ -63,20 +63,32 @@ public CachedDispatchNode(
         this.next = next;
     }
 
+    protected boolean sendingFrames() {
+        return sendsFrame || readCaller != null;
+    }
+
     public void replaceSendingFrame() {
-        CompilerDirectives.transferToInterpreterAndInvalidate();
+        assert readCaller == null;
         if (!sendsFrame) {
-            sendsFrame = true;
+            replaceSendingFrame(true, null);
         }
     }
 
     public void replaceSendingCallerFrame(CallerFrameAccess access) {
-        CompilerDirectives.transferToInterpreterAndInvalidate();
+        assert sendsFrame == false;
         if (readCaller == null) {
-            readCaller = new ReadCallerFrameNode(access);
+            replaceSendingFrame(false, new ReadCallerFrameNode(access));
         }
     }
 
+    private void replaceSendingFrame(boolean sendsFrame, ReadCallerFrameNode readCaller) {
+        CachedDispatchNode copy = (CachedDispatchNode) copy();
```

Comment:
After discussing with Christian Humer, I think the best solution is to use one Assumption per RubyRootNode and check it in the execute() method before executing the child. This limits invalidation to the caller method when discovering the caller frame is needed.